Add bindings for collapser

reds-heig opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the guidelines for contributing: https://giotto-ai.github.io/gtda-docs/latest/contributing/#guidelines
-->

**Reference issues/PRs**
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests.

Pull requests which are not related to open issues may be rejected!
If suggesting a new feature or change, please discuss it in an issue first.
If fixing a bug, there should be an issue describing it with steps to reproduce.
-->

**Types of changes**
<!--
What types of changes does your code introduce? Put an `x` in all the boxes that apply.
-->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

**Description**
<!--
Describe your changes in detail.
-->

Add bindings for collapser from Gudhi new feature.

**Screenshots (if appropriate)**

**Any other comments?**

Currently work in progress, there's a bug on the size of the matrix returns by the C++ in python.

**Checklist**
<!--
Go over all the following points, and put an `x` in all the boxes that apply. 
If you're unsure about any of these, don't hesitate to ask. We're here to help!
-->
- [x] I have read the [guidelines for contributing](https://giotto-ai.github.io/gtda-docs/latest/contributing/#guidelines).
- [x] My code follows the code style of this project. I used `flake8` to check my Python changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed. I used `pytest` to check this on Python tests.

<!--
We value all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.

Thanks for contributing!
-->

Commit:
[WIP] Add bindings for collapser

Currently supports Sparse and Dense representation as input
returns a sparse matrix

Signed-off-by: julian <julian.burellaperez@heig-vd.ch>

ulupo committed:
Attempt fix of collapser bindings

ulupo committed:
Split flag_complex_collapse_edges into a _sparse and a _dense version

ulupo committed:
Clean Cech test docstring

ulupo committed:
Linting in ripser_interface

ulupo committed:
Make more collapser bindings which compile

ulupo committed:
Add collapse_edges kwarg to ripser to enable edge collapser

Commit:
Update collapser bindings

* All functions have the same name, supporting different input types
* All functions returns now tuple<vector, vector, vector>

Signed-off-by: julian <julian.burellaperez@heig-vd.ch>

Commit:
Add test for the collapser bindings

Signed-off-by: julian <julian.burellaperez@heig-vd.ch>

Commit:
Update azure-pipeline ccache to allow building them again

Signed-off-by: julian <julian.burellaperez@heig-vd.ch>

Commit:
Update bindings names and add threshold parameter

Eigen constructor allows vector of vector, this clashes with one of our input
Therefore, we choose to use a different name for each input type

Signed-off-by: julian <julian.burellaperez@heig-vd.ch>

Commit:
Remove unused function from bindings

Signed-off-by: julian <julian.burellaperez@heig-vd.ch>

Commit:
Refactor ripser_interface

Signed-off-by: julian <julian.burellaperez@heig-vd.ch>

Commit:
Add test cases for ripser bindings with collapser

Signed-off-by: julian <julian.burellaperez@heig-vd.ch>

ulupo committed:
Remove unnecessary backslash

ulupo commented on gtda/externals/python/ripser_interface.py:
```diff
@@ -213,29 +213,30 @@ def ripser(X, maxdim=1, thresh=np.inf, coeff=2, metric="euclidean",
+                # If the matrix is already COO, we need to order the row and
+                # column indices lexicographically to avoid errors. See
+                # https://github.com/scikit-tda/ripser.py/issues/103
+                row, col, data = _lexsort_coo_data(dm.row, dm.col, dm.data)
```
With this version, `_lexsort_coo_data` is done twice when the input is coo and `collapse` is `False`. Can we find a way to avoid this?

MonkeyBreaker replied:
We could replace `row, col, data = _lexsort_coo_data(dm.row, dm.col, dm.data)` by `row, col, data = dm.row, dm.col, dm.data`.

I think it would be the easiest solution

ulupo replied:
Yep, I agree!

ulupo replied:
Given what I wrote below, I think it's enough to just leave the lexsort here and remove it from line 236.

MonkeyBreaker replied:
But when collapsing edges is enable, we are sure that we shouldn't sort after collapsing ?

ulupo replied:
Oh you are right! So maybe it should be put inside the `if` branch?

MonkeyBreaker replied:
If I understood you right, only sort if `collapse_edges` is true then ?
Otherwise, when `collapse_edges` is false, it's already sorted because on both cases we have a COO matrix format ?

ulupo replied:
These are the situations in which we need to sort:
- after any collapse (from sparse or dense matrix);
- if the input is COO, regardless of whether we collapse or not.

By the way, I notice that collapse is not yet implemented in the dense case (I thought it was when I first refactored this file).

MonkeyBreaker replied:
If I'm not wrong, this means that we should *not* sort when `collapse_edges` is disable *and* matrix is originally in COO format ?

In this case we should add the sorting when `collapse_edges` is enable or if `collapse_edges` is disable *and* matrix format is originally in COO format. 

I would have like to prevent to duplicate code or add a complicated if before calling the sort function, and sort in all cases, even if in the case where the input matrix is originally in sparse format (but not COO) and it's converted to COO format.

What do you think ?

About the dense case, isn't it covered here ? https://github.com/giotto-ai/giotto-tda/pull/469/files#diff-540ff315655f258b36b02c3a50f10946R216-R218

ulupo replied:
Hmm, I'm not sure why you say that, sorry! I said:

> if the input is COO, regardless of whether we collapse or not.

If we don't collapse, other sparse formats are safe. After we collapse, nothing is safe. But even if we don't collapse, COO needs the sorting.

ulupo commented on gtda/externals/python/ripser_interface.py:
```diff
@@ -206,31 +213,45 @@ def ripser(X, maxdim=1, thresh=np.inf, coeff=2, metric="euclidean",
+                row, col, data = _lexsort_coo_data(dm.row, dm.col, dm.data)
+            else:
+                coo = dm.tocoo()
+                row, col, data = coo.row, coo.col, coo.data
```
I expect that these are automatically lexsorted by scipy when running `tocoo`.

MonkeyBreaker replied:
True

ulupo committed:
Simplify code for upper diagonal extraction in dense case

ulupo commented on gtda/externals/python/ripser_interface.py:
```diff
@@ -206,31 +213,44 @@ def ripser(X, maxdim=1, thresh=np.inf, coeff=2, metric="euclidean",
-            coeff,
+            coeff
             )
     else:
```
This `else` is the dense case. It does not look to me like we use collapse because we just call `DRFDM`, i.e. ripser. What am I missing?

Commit:
Add Flag to indicate when to sort data before computing persistent homology

Signed-off-by: julian <julian.burellaperez@heig-vd.ch>

Commit:
Fix test for ripser

When it's an sparse matrix, it should be upper/lower triangular or
mirrored on the diagonal => i,j == j,i

Signed-off-by: julian <julian.burellaperez@heig-vd.ch>

Commit:
Add threshold directly into the collapser

Signed-off-by: julian <julian.burellaperez@heig-vd.ch>

ulupo committed:
Add references to edge collapse paper

ulupo committed:
Guess number of points using max of shape tuple instead of first shape

Avoids having to reshape sparse arrays explicitly

ulupo committed:
Simplify lexsort flag logic

ulupo committed:
Small refactor of ripser interface to restrict collapse to data without non-zero vertex weights

ulupo committed:
Add hypothesis tests for collapser

Check for barcode equalities

ulupo committed:
Place hypothesis tests in test_ripser and remove superseded tests there. Add regression test for #465

ulupo committed:
Fix array shape

Commit:
Disable computation of cocycles by default

Signed-off-by: julian <julian.burellaperez@heig-vd.ch>
